assignments/introductory-assignments.md

```diff
@@ -488,14 +582,767 @@ A digital circuit computer with 2GB of memory can distinguish 2^{17 179 869 184}
+|       | 1     | 1     | 1     | 1     | 1     | 1     | 1     | 1     |
+| +     | 0     | 0     | 0     | 0     | 0     | 0     | 0     | 1     |
+| **1** | **0** | **0** | **0** | **0** | **0** | **0** | **0** | **0** |
+
+If we have more than 8 bits, this is not a problem. However, with computers, everything is finite, in particular memory. 
+ 
+Moreover, arithmetic operations are on most machines implemented for bit strings with a fixed size, such as 8 bits. On such machines adding 11111111 and 00000001 results in what is called *arithmetic overflow*.
+
+[Arithmetic Overflow](https://en.wikipedia.org/wiki/Arithmetic_overflow): This occurs when an arithmetic operation attempts to create a numeric value too large to be represented within the available storage space.
+
+How can we deal with arithmetic overflow? Two approaches can be combined: *detection* and *semantics*. If an arithmetic overflow is detected, one can discard the computation and do something else. 
+
+For this purpose, most processors feature a so-called carry bit or [carry flag](https://en.wikipedia.org/wiki/Carry_flag). The carry flag is set if an arithmetic operation causes an overflow indicated by a carry out of the most significant bit. In our example, the 9-th bit in 100000000 is that carry bit.
+
+In terms of semantics, if the result of an arithmetic overflow has a defined value, one may be able to use that value in a meaningful way. For example, a common semantics for n-bit arithmetics is to compute everything modulo 2^{n}, also referred to as wrap-around semantics, or wrap around. 
+
+For example, 255+1=256 modulo 2^8=256 modulo 256=0, which is precisely what 100000000 in an 8-bit system stands for. Some applications are correct even when such wrap-around occurs.
+
+[Modulo](https://en.wikipedia.org/wiki/Modulo_operation): The remainder after division of one number by another, also called modulus.
+
+Arithmetic overflow nevertheless is the cause of numerous software bugs and even costly accidents. Restricting the space available for representing something that can be arbitrarily large such as numbers, has serious consequences. Integer arithmetics are always an approximation of real arithmetics. For correctness, computer applications need to be appropriately adapted to work for integer arithmetics, not real arithmetics.
+
 ---
 
-# 2) Architecture
+# Architecture
 
 ## Binary vs. Decimal
 
-Computers encode all information in binary representation, since they also use a binary representation at the hardware level. For example, a 0 is represented by a voltage of -5V or by the state "not electrically charged" and a 1 for example by a voltage +5V or by the state "electrically charged".
+Computers encode all information in binary representation since they also use a binary representation at the hardware level. For example, 0 is represented by a voltage of -5V or the state "not electrically charged" and 1, by a voltage +5V or the state "electrically charged".
```

Comment:
```suggestion
Computers encode all information in binary representation since they also use a binary representation at the hardware level. For example, 0 is represented by a voltage of -5V or the state not electrically charged and 1, by a voltage +5V or the state electrically charged.
```